rosbag2_transport/src/rosbag2_transport/player_service_client.cpp

```diff
@@ -0,0 +1,298 @@
+              RCUTILS_LOG_WARN_ONCE_NAMED(
+                ROSBAG2_TRANSPORT_PACKAGE_NAME,
+                "The configuration of introspection for '%s' is metadata !",
+                service_name_.c_str());
+              return false;
+            }
+          }
+        default:  // request_info_from::SERVICE:
+          {
+            // Already decide using request data from service.
+            return true;
+          }
+      }
+    } else {
+      if (service_set_introspection_content) {
+        request_info[client_id] = request_info_from::SERVICE;
+        return true;
+      } else {
+        request_info[client_id] = request_info_from::NO_CONTENT;  // Only have metadata
+        return false;
+      }
+    }
+  }
+
+  // type is service_msgs::msg::ServiceEventInfo::REQUEST_SENT
+  if (iter != request_info.end()) {
+    switch (iter->second) {
+      case request_info_from::CLIENT:
+        {
+          // Already decide using request data from client.
+          return true;
+        }
+      case request_info_from::NO_CONTENT:
+        {
+          if (rosbag2_cpp::introspection_include_metadata_and_contents(message.size())) {
+            // introspection type is changed from metadata to metadata + contents
+            request_info[client_id] = request_info_from::CLIENT;
+            return true;
+          } else {
+            RCUTILS_LOG_WARN_NAMED(
+              ROSBAG2_TRANSPORT_PACKAGE_NAME,
+              "The configuration of introspection for '%s' client [ID: %s]` is metadata !",
+              rosbag2_cpp::client_id_to_string(client_id).c_str(),
+              service_name_.c_str());
+            return false;
+          }
+        }
+      default:  // request_info_from::SERVICE:
+        {
+          // Already decide using request data from service.
+          return false;
+        }
+    }
+  } else {
+    if (rosbag2_cpp::introspection_include_metadata_and_contents(message.size())) {
+      request_info[client_id] = request_info_from::CLIENT;
+    } else {
+      request_info[client_id] = request_info_from::NO_CONTENT;
+    }
+  }
+
+  return true;
+}
+
+void PlayerServiceClient::async_send_request(const rclcpp::SerializedMessage & message)
+{
+  int ret = RMW_RET_OK;
+
+  {
+    auto ros_message = std::make_unique<uint8_t[]>(message_members_->size_of_);
+
+    message_members_->init_function(
+      ros_message.get(), rosidl_runtime_cpp::MessageInitialization::ZERO);
+
+    ret = rmw_deserialize(&message.get_rcl_serialized_message(), ts_, ros_message.get());
+    if (ret == RMW_RET_OK) {
+      if (client_->service_is_ready()) {
+        // members_[0]: info, members_[1]: request, members_[2]: response
+        auto request_offset = message_members_->members_[1].offset_;
+        auto request_addr = reinterpret_cast<size_t>(ros_message.get()) + request_offset;
+        auto future_and_request_id = client_->async_send_request(
+          reinterpret_cast<void *>(*reinterpret_cast<size_t *>(request_addr)));
```

Comment:
It seems here is a mistake in pointers arithmetic due to the back and force casting with `reinterpret_cast`
It would be much simpler and less error-prone if simplify these 3 lines of code to the
```suggestion
        auto request_offset = message_members_->members_[1].offset_;
        auto future_and_request_id = client_->async_send_request(
          reinterpret_cast<void *>(ros_message.get() + request_offset));
```

Reply:
Original code `reinterpret_cast<void *>(*reinterpret_cast<size_t *>(request_addr)))` means
1. Get the value at the address pointed by request_addr.
2. This value is convert to pointer. 

`reinterpret_cast<void *>(ros_message.get() + request_offset)` is different with original code.